### PR TITLE
TEST: build sass via npm-scripts

### PR DESCRIPTION
**Pull request by bryceosterhaus**
PR to test CI based on npm-script changes from https://github.com/liferay/liferay-frontend-projects/pull/282

Updated `gradle-plugin-css-builder` to only run if the `build` script in the package.json does not exist. This way we can incrementally update modules to build sass via npm-scripts. I'm not totally sure if this is the exact change we would want, but it seemed to work fine locally.

Hoping CI doesn't hate this...

### CODE REVIEW

File: modules/yarn.lock
Comment (jbalsas) [inline modules/yarn.lock:1750]: Do we really need bourbon? I thought we already got rid of it... 🤔

**Review by jbalsas — APPROVED** [review]
LGTM after a very high level overview.

> Updated gradle-plugin-css-builder to only run if the build script in the package.json does not exist. This way we can incrementally update modules to build sass via npm-scripts. I'm not totally sure if this is the exact change we would want, but it seemed to work fine locally.

Sounds like a fair initial step! I think we probably want to just add `package.json` and `build` scripts to every `-web` module that doesn't have it to make sure also other things like minification all go through the same build pipeline.

We'll need to send this through @petershin for him to validate the changes.

**Comment by wincent** [comment]
> I think we probably want to just add `package.json` and `build` scripts to every `-web` module that doesn't have it to make sure also other things like minification all go through the same build pipeline.

This is what I was trying to get at [here](https://github.com/liferay/liferay-frontend-projects/pull/282#pullrequestreview-548946252):

> * There is some value in making _all_ modules have the same build entry point; those modules which don't use `@liferay/npm-scripts` could start saying `liferay-npm-scripts build` instead of `liferay-npm-scripts build-variation-3` (just thinking about a future world where we have `liferay-npm-scripts build-variation-11` etc). Basically, trying to manage the complexity of configuring build tooling _inside_ `@liferay/npm-scripts` intead of strewn across the `package.json` files of the liferay-portal repo; we already have build commands... they grow like weeds.

File: modules/yarn.lock
Comment (bryceosterhaus) [reply modules/yarn.lock:1750]: Yeah its used in a few places and the frontend-css-common module exposes it to the `buildCSS` task to consume in the sass build process

**Comment by bryceosterhaus** [comment]
_its green_

me:



> we probably want to just add package.json and build scripts to every -web module

Yep, my plan is go through and update existing modules now. I wanted to first test this and see if I broke everything first. I'll send another update today to test against CI

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
ci:test:bundle

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
Alright, I have no idea why this isn't working when running tomcat. I'd be stoked if someone could help take a look and see where I am going crazy...

[`ant all` logs](https://gist.github.com/bryceosterhaus/99e45b14d8920a36361c894cab8724a6)
[first startup logs](https://gist.github.com/bryceosterhaus/98ac4516693c27bfc0d03de1dec200e2)
[first page hit logs](https://gist.github.com/bryceosterhaus/36b5fccf3cf289f12ad0952fb600225b)

I've tried not updating the gradle plugins and just having the new scripts coexist with the old build system but that errors as well. I can't seem to pin point why tomcat would blow up with these changes

**Comment by bryceosterhaus** [comment]
ci:test:bundle

**Comment by wincent** [comment]
> I've tried not updating the gradle plugins and just having the new scripts coexist with the old build system but that errors as well. I can't seem to pin point why tomcat would blow up with these changes

I don't really know but:

```
java.lang.IllegalStateException: Unable to find META-INF/resources/package.json in com.liferay.fragment.renderer.react.impl
```

might be the most promising place to go fishing...

`modules/apps/fragment/fragment-renderer-react-impl` has a `package.json` with `"build"` script that uses `liferay-npm-scripts`:

```
        "scripts": {
                "build": "liferay-npm-scripts build",
                "checkFormat": "liferay-npm-scripts check",
                "format": "liferay-npm-scripts fix"
        },
```

but there is nothing under `META-INF/resources` at all (except a `.touch` file to keep what would otherwise be an empty directory tracked in Git):

```
src
└── main
    ├── java
    │   └── com
    │       └── liferay
    │           └── fragment
    │               └── renderer
    │                   └── react
    │                       └── internal
    │                           ├── model
    │                           │   └── listener
    │                           │       └── FragmentEntryLinkModelListener.java
    │                           └── renderer
    │                               └── FragmentEntryFragmentRendererReact.java
    └── resources
        └── META-INF
            └── resources
                └── .touch

14 directories, 3 files
```

I know [you added code to not invoke the bundler](https://github.com/liferay/liferay-frontend-projects/pull/282#issuecomment-745580363) in the absence of JS files, and that seems to be the case for this module, which is probably preventing some necessary manifest file from being created. cc @izaera

**Comment by bryceosterhaus** [comment]
I did a little more testing on this today and got some helpful results.

|  DXP Commit |  npm-scripts version | Builds  | Runs Tomcat  | 
|---|---|---|---|
|  master | 36.2.0 (current DXP)  | ✅   | ✅   | 
|  master |  36.5.1(latest release) |  ✅  | ✅   | 
|  master |  master(latest commit) |  ✅  | ❌ |
|  master |  36.6.0-pre.x |  ✅  | ❌ | 


I think there was something introduced in latest master that isn't released yet that is causing tomcat to blow up. The tomcat logs look something like

```
22-Dec-2020 00:05:21.356 SEVERE [http-nio-8080-exec-1] org.apache.catalina.core.StandardWrapperValve.invoke Servlet.service() for servlet [jsp] in context with path [] threw exception
	java.lang.ArrayIndexOutOfBoundsException: 0
		at com.liferay.portal.util.PortalInstances.getDefaultCompanyId(PortalInstances.java:202)
		at com.liferay.portal.util.PortalImpl.getDefaultCompanyId(PortalImpl.java:2145)
		at com.liferay.portal.kernel.util.PortalUtil.getDefaultCompanyId(PortalUtil.java:866)
		at com.liferay.portal.servlet.filters.validhostname.ValidHostNameFilter.doFilterTry(ValidHostNameFilter.java:39)
		at com.liferay.portal.kernel.servlet.filters.invoker.InvokerFilterChain.processDirectCallFilter(InvokerFilterChain.java:194)
		at com.liferay.portal.kernel.servlet.filters.invoker.InvokerFilterChain.doFilter(InvokerFilterChain.java:99)
		at com.liferay.portal.kernel.servlet.filters.invoker.InvokerFilter.doFilter(InvokerFilter.java:104)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:1
```
 
So I don't actually think it has to do with the sass stuff. I am going to try my commits based off 36.5.1 and see if that works.

**Comment by bryceosterhaus** [comment]
It also could be some issue with using the `install-scripts.sh` with `ant all`. Very inconsistent results on my end...

**Comment by jbalsas** [comment]
Hey @bryceosterhaus, so I've tested this and the errors are indeed caused by [`runBundler`](https://github.com/liferay/liferay-frontend-projects/pull/282/files#diff-11e22f784fbebb6c641bc37725eb1c04950fef432d53284b656efff65ffd34c3R116) not running.

This causes the `package.json` and `manifest.json` files not to be placed inside the `META-INF/resources` folder of your `jar` files.

For example, in the case of the `fragment.renderer.react.impl`, you can verify this by doing:
- `cd tools/sdk/dist`
- `vim com.liferay.fragment.renderer.react.impl-2.0.0.jar`

You should see something like:

Screen Shot 2020-12-22 at 15 57 59

This is the same for all errors, so as soon as one is cleared, another pops up. Stay on the lookout for this part of the message:

> Unable to find META-INF/resources/package.json in com.liferay.frontend.js.collapse.support.web

This brings me to another discovery. There are 2 different issues:

### Modules without JS in `src`
This is the case of `fragment-react-renderer-impl`. It indeed has no JS, but that doesn't mean it doesn't have to build. Other similar modules to this would be `frontend-js-metal-web`. I guess this somehow invalidates the `hasJS` rule as a valid indicator of when the bundler needs to run.

### Modules with JS in `src`
Even in modules with JS in the `src`, like `frontend-js-tooltip-support-web` or `frontend-js-collapse-support-web`, `hasJS` remains false and the bundler fails to run.

That is because the `expandGlobs([path.join(BUILD_CONFIG.input, '**/*.{js,jsx,ts,tsx}')])` always returns an empty array. I don't think the `{}` group in the glob is working as you expect it to. I suspect there's either something in the way we expand the globs but using `**/*.js` does indeed work and makes the bundler run.

I'll keep looking around, but I wanted to make sure I wrote this down so you could have something to look into.

**Comment by bryceosterhaus** [comment]
@jbalsas I'm able to reproduce the no package.json issue even if the bundler runs.

I just tested by removing the `if` block around the bundler and I still get the same error. 🤷‍♂️

> Unable to find META-INF/resources/package.json in com.liferay.fragment.renderer.react.impl

**Comment by jbalsas** [comment]
Are you sure the modules are being built after the changes?

Keep in mind that Gradle will cache inputs/outputs so a change in npm-scripts won't trigger the build tasks in a module. This can be observed looking at the tasks report. You should see them as "Skipped" or "From Cache".

To make sure the module rebuilds, you should use "gradlew deploy -a --rerun-tasks".

Could you double check just to make sure?

**Comment by bryceosterhaus** [comment]
> so a change in npm-scripts won't trigger the build tasks in a module

Even in the case of running `ant all`?

My steps have been
- run install scripts to link npm-scripts to portal
- run `yarn install` from modules
- run `ant all`


> To make sure the module rebuilds, you should use "gradlew deploy -a --rerun-tasks".

Is this before or after running `ant all`? Or should I not be running that at all?

**Comment by jbalsas** [comment]
> Even in the case of running ant all?

In this case, it should, what I'd be worried about in this case is wether the changes you've made to skip the bundle phase are being preserved and not nuked by the `yarnInstall` step that `ant all` will trigger.

> Is this before or after running ant all? Or should I not be running that at all?

Definitely after.  `ant all` should work pretty much in the same way, but because it runs the additional `yarnInstall`, some times makes things harder to test, so I like trying with single modules before going through `ant all`. It could be that there's some additional issue in `ant all` as well too, but since I saw the issue directly in `gradlew deploy -a --rerun-tasks`, that's the first I'd say we should fix.

**Comment by bryceosterhaus** [comment]
@jbalsas thats what I thought. I just released a new preview version of scripts, so I am going to try that with ant all just to be sure what I am seeing is consistent with my local tests.

I'll try `gradlew deploy -a --rerun-tasks` as well

**Comment by bryceosterhaus** [comment]
> I like trying with single modules before going through ant all

That was part of my issue. Building individual modules was working for me, but it was crashing on ant all. So maybe it was the `yarnInstall` that was screwing it up.

**Comment by bryceosterhaus** [comment]
Latest pre-release works for me. Meaning that a lot of my issues have likely stemmed from using `install-scripts.sh` paired with `ant all`... Hopefully should get this to work soon then

**Comment by bryceosterhaus** [comment]
Tried running locally with `gradlew deploy -a --rerun-tasks ` and it fails with...

> Task :core:osgi-service-tracker-collections-test:compileTestIntegrationJava
/Users/bryceosterhaus/repos/liferay-portal/modules/core/osgi-service-tracker-collections-test/src/testIntegration/java/com/liferay/osgi/service/tracker/collections/map/test/ListServiceTrackerMapTest.java:17: error: package com.liferay.arquillian.extension.junit.bridge.junit does not exist
import com.liferay.arquillian.extension.junit.bridge.junit.Arquillian;
                                                          ^

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
I think this is interacting with https://github.com/liferay-frontend/liferay-portal/pull/606

I'm seeing:

> index.es.js?languageId=en_US:4 Uncaught TypeError: getModule is not a function
>    at index.es.js?languageId=en_US:4
>    at index.es.js?languageId=en_US:14

Which I guess can be explained by that. @izaera, can you confirm? I can push your changes on top of this one to verify.

**Comment by izaera** [comment]
That is really strange, because the theory is that my changes do nothing unless you have a `federation` entry in your `npmscripts.config.js`...

Also #606 is not merged yet. How could that interact with this? :thinking:

**Comment by izaera** [comment]
I will download this PR to my local env and build it :shrug:

**Comment by wincent** [comment]
> That is because the `expandGlobs([path.join(BUILD_CONFIG.input, '**/*.{js,jsx,ts,tsx}')])` always returns an empty array. I don't think the `{}` group in the glob is working as you expect it to. I suspect there's either something in the way we expand the globs but using `**/*.js` does indeed work and makes the bundler run.

That's correct. `expandGlobs()` just processes very simple globs (which most of our globs are). If you want handling of `{a,b,c}` type patterns you use `preprocessGlobs()` ([behavior of that one can be seen here](https://github.com/liferay/liferay-frontend-projects/blob/542e42c5b24897a1e63016afae7c44e7351e526e/projects/npm-tools/packages/npm-scripts/test/utils/preprocessGlob.js) to effectively "explode" compound globs into a set of simple non-compound globs.

Why is it split into different functions like that? Part of it is [historical](https://github.com/liferay/liferay-frontend-projects/commit/50c7d8c2e7e3a9b74baca4caf470d388554dcd72), and the other is that we don't often need to deal with complicated globs using these low-level functions directly; our main use case was the high-level abstraction of [`getPaths()`](https://github.com/liferay/liferay-frontend-projects/blob/master/projects/npm-tools/packages/npm-scripts/src/utils/getPaths.js), which handles more "stuff" was relevant when we did all this for linting and formatting:

- Respecting ignore files like `.eslintignore` and `.prettierignore`.
- Calling `preprocessGlobs()` to deal with compound globs.
- Calling `filterGlobs()` to filter out globs which don't apply to particular file types.
- Actually calling `expandGlobs()`.
- Filtering out files that are changed on the current branch.

This may all seem complicated, but the design (ie. composing very simple functions together) made sense for the initial use case. I think, now, that the evidently-not-obvious-enough behavior of `expandGlobs()` (specifically, what it _doesn't_ do) suggest that we might want to revisit this and make it at least call `preprocessGlobs()`. We'd lose some of the simplicity/composibility of the individual functions and increase the coupling, but it would make it easier to use.

**Comment by wincent** [comment]
> I think, now, that the evidently-not-obvious-enough behavior of `expandGlobs()` (specifically, what it _doesn't_ do) suggest that we might want to revisit this and make it at least call `preprocessGlobs()`.

Heh, or now that I think of it. Just use `getPaths()`? (Or if there is something making it unsuitable for use, make it suitable.)

**Comment by izaera** [comment]
I can confirm this is broken (probably due to my changes to `npm-scripts`).

**Comment by izaera** [comment]
> That is really strange, because the theory is that my changes do nothing unless you have a federation entry in your npmscripts.config.js...

Ah, OK, now I understand. We already merged two modules with `federation` set to `true` and because this PR is getting some intermediate version for npm-scripts it breaks. Let's see if updating to latest `npm-scripts` that I released this morning fixes it.

Will report in a while...

**Comment by izaera** [comment]
Of course not, because we need changes in `liferay-portal` to adapt to latest `npm-scripts` version :tada: 

I would say we can rebase this on top of #606 to get the server changes and everything should work. #606 is already green, so it should be merged soon.

**Comment by izaera** [comment]
I have rebased (need to skip the two commits updating `npm-scripts` version). I will `ant all` now.

**Comment by izaera** [comment]
After ant all this is still broken but for different reasons.

**Comment by jbalsas** [comment]
> After ant all this is still broken but for different reasons.

Which reasons, we might never know... :joy:

Portal looked visually okay for me as far as CSS went. Couldn't navigate anywhere, though, which could account for the errors we're seeing in CI :( 

I won't have time to give this another try on top of @izaera's PR, so maybe we'll have to get back to this next week 😢 

@bryceosterhaus, one interesting thing to do would be to also measure compile time with and without this to compare before and after. Brian (and ourselves) will want this information :)

**Comment by bryceosterhaus** [comment]
> measure compile time with and without this to compare before and after

@jbalsas checking that out today!

**Comment by bryceosterhaus** [comment]
So I ran some tests across a few random modules with various amounts of scss. In order to time to buildSass task from npm-scripts, I purposely isolated it to only test the time for sass and not the whole build script.

| Module               | buildCSS (gradle) | buildSass (npm-scripts) |
|----------------------|-------------------|-------------------------|
| `app-builder-web`      | 2s                | 1.06s 🏎️                   |
| `document-library-web` | 6s                | 0.69s 🏎️                    |
| `fragment-web`        | 6s                | 4.04s   🏎️                 |
| `segments-web`         | 2s                | 0.8s    🏎️                 |

The speed of npm-scripts was favored in every module I tested, so that seems like a good sign here.

**Comment by jbalsas** [comment]
> The speed of npm-scripts was favored in every module I tested, so that seems like a good sign here.

That's encouraging! /cc @david-truong, @petershin

**Comment by jbalsas** [comment]
@bryceosterhaus, I think this will still be blocked by @izaera's other PR. One additional thing I think we should do would be to add automatically the `clay` paths to the `includePath` list in `npm-scripts`. That way we can remove the unnecessary configuration in `build.gradle`. 

We just kept it that way because the gradle script used to add a different set of `includePaths` and didn't want to make a breaking change there. Since this is a new tool, I think we can afford it and better now than later ;)

**Comment by izaera** [comment]
@jbalsas and @bryceosterhaus can this wait for the other PR to be merged (remember we are waiting for the webpack release) or you prefer that I add some code in npm-scripts to avoid this problem?

**Comment by jbalsas** [comment]
> @jbalsas and @bryceosterhaus can this wait for the other PR to be merged (remember we are waiting for the webpack release) or you prefer that I add some code in npm-scripts to avoid this problem?

I think being Christmas and all we can wait a little... but we should have a plan for when you go on PTO to make sure we don't remain blocked after that ;)

**Comment by wincent** [comment]
> The speed of npm-scripts was favored in every module I tested, so that seems like a good sign here.

Pleasantly surprising, given that one of the major issues with the new Sass implementation [has been the perf, specially on Node](https://github.com/sass/dart-sass/blob/master/perf.md):

> Dart Sass on Node is still substantially slower than on the Dart VM

Basically, the perf of native Dart is pretty good at this point, but the Node version is still slow. Numbers vary depending on the test, but an example result from the linked page is:

> Based on these numbers, Dart Sass from a native executable is approximately:
>
> - 3.8x slower than libsass
> - 10.3x faster than Dart Sass on Node

From which I conclude that there must be some ghastly overhead in calling all this from Java, or our Sass is too puny for any of this to actually matter. Good news in any case, I guess!

**Comment by jbalsas** [comment]
> Pleasantly surprising, given that one of the major issues with the new Sass implementation has been the perf, specially on Node:
>
>Dart Sass on Node is still substantially slower than on the Dart VM

There's also this:

> When installed via npm, Dart Sass supports a JavaScript API that aims to be compatible with Node Sass. Full compatibility is a work in progress, but Dart Sass currently supports the render() and renderSync() functions. Note however that by default, renderSync() is more than twice as fast as render(), due to the overhead of asynchronous callbacks.
>
> To avoid this performance hit, render() can use the fibers package to call asynchronous importers from the synchronous code path. To enable this, pass the Fiber class to the fiber option:

So maybe there's still even some additional room for improvement if we're using the async version.

**Comment by izaera** [comment]
> I think being Christmas and all we can wait a little... but we should have a plan for when you go on PTO to make sure we don't remain blocked after that ;)

I'm exploring the mini-PR now. It should shed some light about this error, serves as a backup plan, and should make testing this easier. I'll report when I have it so that we can rebase this on top of it instead of the whole LPS-124287 PR.

**Comment by izaera** [comment]
I think I've found a workaround. It involves a very small PR in liferay-portal to hide `Liferay.require` under a Symbol-indexed property in `window` and a new release of `npm-scripts` that "reverts" all changes without having to remove the code (in fact, it enhances the tool with more versatility :-D).

I will prepare and send the PRs tomorrow morning so that we have everything ready before lunch (hopefully :crossed_fingers: ). Then we can rebase this PR on top of the one I'll send for portal and everything should work (at least from the bundling point of view).

Then, I will prepare two more PRs so that when webpack is released we can merge them and make everything work again.

**Comment by david-truong** [comment]
@bryceosterhaus buildCSS also does the rtl conversion.  Does your buildSass task also do that?  If I remember correctly... that rtl conversion in java was very slow.  Also another area where the java buildCSS was very slow was when any clay.css needed to be imported...  those might be good places to also test for comparisons

**Comment by bryceosterhaus** [comment]
@david-truong we haven't added rtl at build time yet but we do have a mechanism for it. From what I understand with the java implementation, we are running the rtl conversion at run time and not building it ahead of time. Am I correct in that? My java isn't the sharpest, but from what I was reading through, thats what I had concluded.


And yes, this task is taking clay imports into account.

**Comment by david-truong** [comment]
I think there are mechanisms in portal to generate it during runtime if we it missing but we are still generating them on build time for modules we ship.

**Comment by jbalsas** [comment]
> I think there are mechanisms in portal to generate it during runtime if we it missing but we are still generating them on build time for modules we ship.

I thought so too, but doing a quick check I'm not so sure now 😂

**Comment by izaera** [comment]
@bryceosterhaus (et al.) It looks like rebasing this on top of https://github.com/liferay-frontend/liferay-portal/pull/660 makes it build and work (at least I don't see error in the server or browser consoles after a quick navigation).

Given that I cannot push to `bryceosterhaus/liferay-portal`, I've left the rebase in a [branch of mine](https://github.com/izaera/liferay-portal/commits/248). It's just this PR minus the "update npm scripts" commits which conflict with #660 and are not needed.

**Comment by bryceosterhaus** [comment]
> I think there are mechanisms in portal to generate it during runtime if we it missing but we are still generating them on build time for modules we ship.

@david-truong @jbalsas if you find an example of this, could you send it my way? The only thing I saw related to RTL conversion in our buildCSS gradle task is the ability to add an "excludes" pattern.

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
Most recent updates from https://github.com/liferay/liferay-frontend-projects/pull/282 being tested with a pre-release here...

**Comment by john-co** [comment]
Reviewed. Test failures are unrelated. :heavy_check_mark: 

https://testray.liferay.com/home/-/testray/runs/compare?formDate=1609875370756&testrayRunIdA=1389448867&testrayRunIdB=1389618298

modules-unit-jdk8/0 test failures may be a fluke, but I can't tell off the bat. Recommend to rerun the test suite to confirm.

**Comment by bryceosterhaus** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
rebased and re-built gradle plugins

**Comment by john-co** [comment]
Reviewed. Test failures are unrelated. :heavy_check_mark: 
- Forms tests are flaky, also observed in other PR https://github.com/liferay-frontend/liferay-portal/pull/618#issuecomment-754927319

Ok to manually forward.

**Comment by bryceosterhaus** [comment]
closing for now. Will re-open when there is an official release of npm-scripts

**Pull request closed without merging**
